### PR TITLE
wallet-ext: initial api client for qredo

### PR DESCRIPTION
**Pull request by pchrysochoidis**
part of APPS-760

### CODE REVIEW

File: apps/wallet/src/shared/qredo-api.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    public getWallets({
+        filters,
+    }: GetWalletsParams = {}): Promise<GetWalletsResponse> {
+        const searchParams = new URLSearchParams();
+        if (filters?.address) {
+            searchParams.append('address', filters.address);
+        }
+        const searchQuery = searchParams.toString();
```
Comment (williamrobertson13) [inline apps/wallet/src/shared/qredo-api.ts:111]: super nit: I think `searchQuery` will automatically be converted to a string when used in the template string so you could probs delete this line and just reference `searchParams` directly

File: apps/wallet/src/shared/qredo-api.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+export type QredoAPIErrorResponse = {
+    code: string;
+    msg: string;
+    detail: {
+        reason: string;
+    };
+};
+
+export class QredoAPIError extends Error {
```
Comment (williamrobertson13) [inline apps/wallet/src/shared/qredo-api.ts:12]: just an FYI: you might have to do some weird stuff if you want to be able to do `instanceof` checks with these custom exceptions: https://stackoverflow.com/questions/31626231/custom-error-class-in-typescript

File: apps/wallet/src/shared/qredo-api.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            `${this.baseURL}connect/sui/wallets${
+                searchQuery ? `?${searchQuery}` : ''
+            }`
+        );
+    }
+
+    #request = async (...params: Parameters<typeof fetch>) => {
+        let tries = 0;
+        while (tries++ <= 1) {
+            // TODO: add monitoring?
```
Comment (williamrobertson13) [inline apps/wallet/src/shared/qredo-api.ts:122]: do we have plans to address this before releasing the Qredo integration?

File: apps/wallet/src/shared/qredo-api.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                    } else {
+                        this.#accessTokenRenewInProgress =
+                            this.#renewAccessTokenFN(this.qredoID).finally(
+                                () => (this.#accessTokenRenewInProgress = null)
+                            );
+                        this.#accessToken = await this
+                            .#accessTokenRenewInProgress;
+                    }
+                    if (this.#accessToken) {
+                        continue;
```
Comment (williamrobertson13) [inline apps/wallet/src/shared/qredo-api.ts:147]: this control flow logic is a tad hard to follow (especially with all the nesting), I wonder if there's a simpler way to re-write this retry / refresh logic?

File: apps/wallet/src/shared/qredo-api.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        return this.#request(
+            `${this.baseURL}connect/sui/wallets${
+                searchQuery ? `?${searchQuery}` : ''
+            }`
+        );
+    }
+
+    #request = async (...params: Parameters<typeof fetch>) => {
+        let tries = 0;
+        while (tries++ <= 1) {
```
Comment (williamrobertson13) [inline apps/wallet/src/shared/qredo-api.ts:121]: nit: it might be nice to make `maxRetries` an optional arg that defaults to 1 so we don't have to reference `1` in multiple places which is slightly error prone

**Review by williamrobertson13 — APPROVED** [review]

File: apps/wallet/src/shared/qredo-api.ts (hunk 2 above)
Comment (Jordan-Mysten) [reply apps/wallet/src/shared/qredo-api.ts:12]: I've heard this a lot but I haven't actually seen this be an issue in the latest browsers?

Comment (pchrysochoidis) [reply apps/wallet/src/shared/qredo-api.ts:12]: It was done like this to make more detailed error, it's not used with instance of at the moment. Will leave it as is for now and revisit in case it creates any problems

File: apps/wallet/src/shared/qredo-api.ts (hunk 1 above)
Comment (pchrysochoidis) [reply apps/wallet/src/shared/qredo-api.ts:111]: The problem is the `?`, we need to append it ourselves
